yt/frontends/amrvac/data_structures.py

```diff
@@ -0,0 +1,255 @@
+        self.num_grids = self.dataset.parameters['nleafs']
+
+    def _parse_index(self):
+        """Populate self.grid_* attributes from tree info from datfile header."""
+        # required method
+        with open(self.index_filename, "rb") as istream:
+            vaclevels, morton_indices, block_offsets = get_tree_info(istream)
+            assert len(vaclevels) == len(morton_indices) == len(block_offsets) == self.num_grids
+
+        self.block_offsets = block_offsets
+        # YT uses 0-based grid indexing, lowest level = 0 (AMRVAC uses 1 for lowest level)
+        ytlevels = np.array(vaclevels, dtype="int32") - 1
+        self.grid_levels.flat[:] = ytlevels
+        self.max_level = np.max(ytlevels)
+        assert self.max_level == self.dataset.parameters["levmax"] - 1
+
+        # some aliases for left/right edges computation in the coming loop
+        domain_width = self.dataset.parameters["xmax"] - self.dataset.parameters["xmin"]
+        block_nx = self.dataset.parameters["block_nx"]
+        xmin = self.dataset.parameters["xmin"]
+        dx0 = domain_width / self.dataset.parameters["domain_nx"] # dx at coarsest grid level (YT level 0)
+        dim = self.dataset.dimensionality
+
+        self.grids = np.empty(self.num_grids, dtype='object')
+        for igrid, (ytlevel, morton_index) in enumerate(zip(ytlevels, morton_indices)):
+            dx = dx0 / self.dataset.refine_by**ytlevel
+            left_edge = xmin + (morton_index-1) * block_nx * dx
+
+            # edges and dimensions are filled in a dimensionality-agnostic way
+            self.grid_left_edge[igrid, :dim] = left_edge
+            self.grid_right_edge[igrid, :dim] = left_edge + block_nx * dx
+            self.grid_dimensions[igrid, :dim] = block_nx
+            self.grids[igrid] = self.grid(igrid, self, ytlevels[igrid])
+
+    def _populate_grid_objects(self):
+        # required method
+        for g in self.grids:
+            g._prepare_grid()
+            g._setup_dx()
+
+
+
+class AMRVACDataset(Dataset):
+    _index_class = AMRVACHierarchy
+    _field_info_class = AMRVACFieldInfo
+
+    def __init__(self, filename, dataset_type='amrvac',
+                units_override=None, unit_system="cgs",
+                geometry_override=None):
+        # note: geometry_override is specific to this frontend
+
+        self._geometry_override = geometry_override
+        super(AMRVACDataset, self).__init__(filename, dataset_type,
+                                            units_override=units_override, unit_system=unit_system)
+        self.fluid_types += ('amrvac',)
+        # refinement factor between a grid and its subgrid
+        self.refine_by = 2
+
+    @classmethod
+    def _is_valid(self, *args, **kwargs):
+        """At load time, check whether data is recognized as AMRVAC formatted."""
+        # required class method
+        validation = False
+        try:
+            with open(args[0], "rb") as istream:
+                get_header(istream)
```

Comment:
What you're doing here is very unsafe. You're not actually checking if that's AMRVAC format, but instead you're trying to parse the file as if it *were* AMRVAC file. The only check you have there is:

```
if struct.unpack(fmt, istream.read(struct.calcsize(fmt)) < 3:
    raise IOError("Unsupported AMRVAC .dat file version: %d", h['datfile_version'])
```

Apparently reading a random bit from a file results in a number less than 3 more often than not. It causes allocation of <insert random big int> few lines later.

Reply:
wow, I can't believe I let that as it is, thank you so much for pointing this out !

Reply:
7e3a432 adresses this problem but is probably a bit too optimistic.